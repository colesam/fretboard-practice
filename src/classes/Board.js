import { NOTES } from '@/globals'
import BoardValidator from '@/validation/Board'

/**
 * Class representing a single boardState object with extra getter methods
 * @property {String} this.id - Unique identifier for the fretboard
 * @property {String} this.name - Name of the fretboard
 * @property {Position[]} this.positions - Enabled positions of the fretboard
 * @property {String} this.root - Root note of the fretboard
 * @property {Array} this.openTuning - Tuning of each note in open position
 * @property {String} this.notePreferences - Preferred names of each note in the chromatic scale
 * @property {Number} this.numFrets - Number of frets on this fretboard
 * @property {Number} this.numStrings - Number of strings on this fretboard
 * @property {Number} this.startingFret - Fret number that the board starts at
 */
class Board {
  /**
   * @param {BoardState} boardState - object representing the state of the board
   */
  constructor(boardState) {
    BoardValidator.validateBoard(boardState)
    this.id = boardState.id
    this.name = boardState.name
    this.positions = boardState.positions
    this.root = boardState.root
    this.openTuning = boardState.openTuning
    this.notePreferences = boardState.notePreferences
    this.numFrets = boardState.numFrets
    this.numStrings = boardState.numStrings
    this.startingFret = boardState.startingFret

    // Extra variable for caching purposes
    this.noteMatrix = null
  }

  /**
   *
   * @returns {BoardState}
   */
  getBoardState() {
    return {
      id: this.id,
      name: this.name,
      positions: this.positions,
      root: this.root,
      openTuning: this.openTuning,
      notePreferences: this.notePreferences,
      numFrets: this.numFrets,
      numStrings: this.numStrings,
      startingFret: this.startingFret
    }
  }

  /**
   * Determines if a position is included in a board's state
   * @param {Position} searchPosition - Position object being searched for
   * @returns {boolean}
   */
  includesPosition(searchPosition) {
    BoardValidator.validatePosition(searchPosition, this.getBoardState())
    const { fret, string } = searchPosition
    return (
      this.positions.find(position => {
        return position.fret === fret && position.string === string
      }) !== undefined
    )
  }

  /**
   * Gets the note at a given position
   * @param {Position} position - Position object used to find the note
   */
  getNote(position) {
    BoardValidator.validatePosition(position, this.getBoardState())
    const { fret, string } = position
    const baseNote = this.openTuning[string]
    const startIndex = NOTES.indexOf(baseNote)
    return NOTES[(startIndex + fret + this.startingFret) % 12]
  }

  /**
   * Gets the preferred notes in the form of an array
   * @returns {Array} preferredNotes
   */
  getPreferredNotes() {
    return Object.keys(this.notePreferences).map(key => this.notePreferences[key])
  }

  /**
   * Gets a 2D array representing the notes at every position on the fretboard
   * @return {Array}
   */
  getNoteMatrix() {
    if (this.noteMatrix !== null) {
      return this.noteMatrix
    } else {
      let matrix = []
      for (let fret = 0; fret < this.numFrets; fret++) {
        matrix.push([])
        for (let string = 0; string < this.numStrings; string++) {
          matrix[fret].push(this.getNote({ fret, string }))
        }
      }
      return matrix
    }
  }
}

/**
 * @typedef {Object} BoardState
 * @property {String} id - Unique identifier for the fretboard
 * @property {String} name - Name of the fretboard
 * @property {Position[]} positions - Enabled positions of the fretboard
 * @property {String} root - Root note of the fretboard
 * @property {Array} openTuning - Tuning of each note in open position
 * @property {Object} notePreferences - Preferred names of each note in the chromatic scale
 * @property {Number} numFrets - Number of frets on this fretboard
 * @property {Number} numStrings - Number of strings on this fretboard
 * @property {Number} startingFret - Fret number that the board starts at
 */

/**
 * @typedef {Object} Position
 * @property {Number} fret - Fret position offset from 0
 * @property {Number} string - String position offset from 0
 */

export default Board
