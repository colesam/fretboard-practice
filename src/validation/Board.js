import { NOTES } from '@/globals'
import ow from 'ow'

/**
 * @namespace Validation
 */

/**
 * Validation module for BoardState objects and properties
 * @namespace Validation.Board
 */
export default {
  /**
   * Returns an object of predicates that validate each property of a BoardState object
   * @memberof Validation.Board
   * @param {BoardState} board
   */
  predicates(board) {
    let position = ow.object
    let openTuning = ow.array.ofType(ow.string.oneOf(NOTES))
    let note = ow.string.oneOf(NOTES)

    // Add additional validation rules that depend on board data
    if (board != null) {
      position = position.exactShape({
        fret: ow.number.inRange(0, board.numFrets),
        string: ow.number.inRange(0, board.numStrings - 1)
      })
      openTuning = openTuning.length(board.numStrings)
    }

    return {
      id: ow.string.nonEmpty,
      name: ow.string.nonEmpty,
      positions: ow.array.ofType(position),
      root: note,
      notePreferences: ow.object.hasKeys(...NOTES),
      numFrets: ow.number.greaterThanOrEqual(0),
      numStrings: ow.number.greaterThanOrEqual(0),
      openTuning,
      note,
      position
    }
  },

  /**
   * Throws ArgumentError if board data is invalid
   * @memberof Validation.Board
   * @param {BoardState} board - Board data to validate
   */
  validateBoard(board) {
    const {
      id,
      name,
      positions,
      root,
      openTuning,
      notePreferences,
      numFrets,
      numStrings
    } = this.predicates(board)

    ow(
      board,
      ow.object.exactShape({
        id,
        name,
        positions,
        root,
        openTuning,
        notePreferences,
        numFrets,
        numStrings
      })
    )
  },

  /**
   * Throws ArgumentError if board name is invalid
   * @memberof Validation.Board
   * @param {string} name - Board name to validate
   */
  validateName(name) {
    ow(name, this.predicates().name)
  },

  /**
   * Throws ArgumentError if note is invalid
   * @memberof Validation.Board
   * @param {string} note
   */
  validateNote(note) {
    ow(note, this.predicates().note)
  },

  /**
   * Throws ArgumentError if position is invalid
   * @memberof Validation.Board
   * @param {Position} position - Position object to validate
   * @param {BoardState} board - Board data used to determine validity
   */
  validatePosition(position, board) {
    ow(position, this.predicates(board).position)
  }
}

// Related Type Definitions
/**
 * @typedef {Object} BoardState
 * @property {string} id - Unique identifier for the fretboard
 * @property {string} name - Name of the fretboard
 * @property {Position[]} positions - Enabled positions of the fretboard
 * @property {string} root - Root note of the fretboard
 * @property {Array} openTuning - Tuning of each note in open position
 * @property {string} notePreferences - Preferred names of each note in the chromatic scale
 * @property {string} numFrets - Number of frets on this fretboard
 * @property {string} numStrings - Number of strings on this fretboard
 */

/**
 * @typedef {Object} Position
 * @property {number} fret - Fret position offset from 0
 * @property {number} string - String position offset from 0
 */
