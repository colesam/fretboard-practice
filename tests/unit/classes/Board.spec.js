import state from '@/store/testData/testState'
import Board from '@/classes/Board'

let board
beforeEach(() => {
  board = new Board(state.boards['1'])
})

describe('includesPosition', () => {
  it('returns true if the position is included in the board', () => {
    expect(board.includesPosition({ fret: 3, string: 0 })).toBe(true)
  })

  it('returns false if the position is not included in the board', () => {
    expect(board.includesPosition({ fret: 19, string: 3 })).toBe(false)
  })
})

describe('getNote', () => {
  it('returns the correct note at the passed position', () => {
    expect(board.getNote({ fret: 0, string: 0 })).toBe('E')
    expect(board.getNote({ fret: 19, string: 3 })).toBe('D')
    expect(board.getNote({ fret: 3, string: 2 })).toBe('F')
    expect(board.getNote({ fret: 4, string: 2 })).toBe('F#')
  })

  it('returns the correct note on boards with a startIndex greater than 0', () => {
    board = new Board(state.boards['3'])
    expect(board.getNote({ fret: 0, string: 0 })).toBe('B')
    expect(board.getNote({ fret: 1, string: 1 })).toBe('F')
    expect(board.getNote({ fret: 2, string: 2 })).toBe('B')
    expect(board.getNote({ fret: 3, string: 3 })).toBe('F')
  })
})
