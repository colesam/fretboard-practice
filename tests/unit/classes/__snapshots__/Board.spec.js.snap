// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getNoteMatrix returns a matrix of every note at every position on the fretboard 1`] = `
Array [
  Array [
    "E",
    "A",
    "D",
    "G",
  ],
  Array [
    "F",
    "A#",
    "D#",
    "G#",
  ],
  Array [
    "F#",
    "B",
    "E",
    "A",
  ],
  Array [
    "G",
    "C",
    "F",
    "A#",
  ],
  Array [
    "G#",
    "C#",
    "F#",
    "B",
  ],
  Array [
    "A",
    "D",
    "G",
    "C",
  ],
  Array [
    "A#",
    "D#",
    "G#",
    "C#",
  ],
  Array [
    "B",
    "E",
    "A",
    "D",
  ],
  Array [
    "C",
    "F",
    "A#",
    "D#",
  ],
  Array [
    "C#",
    "F#",
    "B",
    "E",
  ],
  Array [
    "D",
    "G",
    "C",
    "F",
  ],
  Array [
    "D#",
    "G#",
    "C#",
    "F#",
  ],
  Array [
    "E",
    "A",
    "D",
    "G",
  ],
  Array [
    "F",
    "A#",
    "D#",
    "G#",
  ],
  Array [
    "F#",
    "B",
    "E",
    "A",
  ],
  Array [
    "G",
    "C",
    "F",
    "A#",
  ],
  Array [
    "G#",
    "C#",
    "F#",
    "B",
  ],
  Array [
    "A",
    "D",
    "G",
    "C",
  ],
  Array [
    "A#",
    "D#",
    "G#",
    "C#",
  ],
]
`;

exports[`getNoteMatrix returns a matrix of every note at every position on the fretboard 2`] = `
Array [
  Array [
    "B",
    "E",
    "A",
    "D",
  ],
  Array [
    "C",
    "F",
    "A#",
    "D#",
  ],
  Array [
    "C#",
    "F#",
    "B",
    "E",
  ],
  Array [
    "D",
    "G",
    "C",
    "F",
  ],
  Array [
    "D#",
    "G#",
    "C#",
    "F#",
  ],
  Array [
    "E",
    "A",
    "D",
    "G",
  ],
  Array [
    "F",
    "A#",
    "D#",
    "G#",
  ],
]
`;
